### ISSUE TITLE
Update IMPS 3 for field-testing by other leaders. 

### ISSUE THREAD

**Issue by joereddington** [eQualityTime/Public#152]
IMPS 3 has been successfully piloted by Joe, but it needs work to be reliably used by other leaders. 

# Resources
* #139 was the first moving of the stack. 
* #150 was my test version 

## Etherpad
- [x] Remove keyboard shortcuts for syntax highlighting and italics and so on. 
- [x] Install a spellchecker for Etherpad 
- [x] Consider a syntax highlighter for comments 

## Config
- [x] Hide the 'python' (and other working directories) from public view 

# Good Code 
- [x] Store the code somewhere(!) it's all on the same drive.  
- [x] Record videos showing how things should work. 
- [x] Refactoring for tests
- [x] Working locally
- [x]  the book database should be formated json - not on one line

# Character Pages 
- [x]  mock up a characters.json for horror2022 
- [x]  Run a php server on main machine
- [x]  get php to read the characters.json and provide links. 

# PDFs 
- [x] Write the button that runs the 'make_pdf' script', we'll make it better soon. 
- [x] Into 'update_books' probably 
- [x] Import the 'list files in directory' code.   
- [x] Connect the button and the script

# Central Listing
- [x] Mock up a central listing page 
- [x] Get php to read it, and have switches for active and not (we'll always be running update, even if we aren't running latex) 

# For the next issue
Writers can press a button to get the pdf for their character. 
- [ ] Strengthen the chapter tests 
- [x] Timestamps on 'last updated' 
- [x] Epub files
- [x] index.php needs to have a good display for when it's not active (maybe put a picture in) 
- [ ] Etherpad should had a logo 
- [x] Dashboard should respond to status - only allow pdfs if the right status is choosen
- [ ] Generation of epubs when correct status is chosen
- [ ] *write the document files to the latex sources directory NOT the python one
- [ ] Upgrade the php to latest version and consider phpunittest
- [ ] Bring in the pdf naming tools to the python. Break it down and make it testable. 
- [x] Upgrade the make-snapshot.sh script . It currently takes literally all of the source files every time.  It should be restricted to those camps that are active. 
- [x] functions to safely remove old pads and old source files. 
- [x] In documentation: make clear the tests that are used for generating characters
- [x] tidy up the results of the 'make pdf' php - use the trace.php from

**Comment by joereddington** [comment]
Recorded a video at https://youtu.be/mTzq3EG0DP0

**Comment by joereddington** [comment]
Did some refactoring today and pushed. There's a reasonable list of todos to work on in the code but the next step is definately to do some config work on the etherpad

**Comment by joereddington** [comment]
## 02/03/22 13:05 to 16:10, Etherpad  +EQT

# Etherpad
* Fixed the ssl - turns out I needed to restart nginx as well... 
* Etherpad - this page gave me a command to test in-browser spell checking: https://bugzilla.mozilla.org/show_bug.cgi?id=1218899 
* Syntax highlighting appears to be too difficult, although possible for the future with some javascript. 
* Turned off the keyboard shortcuts (it was in both settings.json and in a javascript file) 

## Rebooting Notes 
Having edited settings.json I needed to restart
* Restarting with normal restart command had NO effect. 
* Restarting by 'stop'ing server and running /src/bin/run.sh did work. 
  * The database stayed were it was though. 

# Python
* Writing tests and code for a sprints database. 
  * Standard thing about the tests forcing me to write much more serious and careful code than otherwise. 
* Have made the json more readable 
* Sprints.json is writing
* Character.json is writing.
* I'm quite aware that the python is lovely and thoroughly tested and the php is a mess that will go wrong. 
* Should I spend a few hours with a book? 
  * Not yet, but it's interesting to consider for the future.

**Comment by joereddington** [comment]
IMPS 3 has gone through another stage of development. 

It's now at the point where is is ready for a field test from someone other than me.  

I've added a bunch of features that make it easier to use: 
* Spellchecking
* Writers can find links to their character dashboards from main dashboard 
* Leaders can now generate pdfs themselves on demand.  

...along with some behind the scenes safety features.

Both of the bottom two features can been seen at: https://whitewaterwriters.com/oracle/index.php?slug=horror2022, which is my testing file. 

The camp management page is now https://whitewaterwriters.com/oracle/germinate.php - this takes over from the old driver function.  You saw the germination processes itself in the video, but I've now added status selectors for the camps. The backend isn't fully connected but we'll shortly be in a position where backups won't be taken unless the camp is active, and final ebook files won't be generated until the camp is marked as publishing (This is mostly to cut down on server usage). 

On pdfs: the only way to generate PDFs in IMPS 3 will be by pressing a button - they won't auto-generated any more (partly for server reasons). I've been able to do this mostly because the new code is *so* much faster.  This also means that there isn't a link to the trace file any more - it turns up when you press the button.  The next step for this is that writers will shortly be able to generate their own pdfs covering only their own chapters.  

I've been quite restrained with this design iteration - the 'for the next issue' section of https://github.com/eQualityTime/Public/issues/152 covers all the bits that I would normally have got distracted with, but will be in future work.
